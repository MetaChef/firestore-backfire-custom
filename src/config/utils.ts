import { cosmiconfig } from "cosmiconfig";

import { FirestoreConnectionOptions } from "~/firestore";
import { ActionOptions } from "~/actions/types";
import { DataSourceOptions } from "~/data-source/interface";

import { NAME } from "./constants";

export type GlobalOptions = {
  /**
   * Path to the configuration file to use.
   */
  config?: string;
};

export type ResolvedConfig = {
  connection: FirestoreConnectionOptions;
  action: ActionOptions;
  dataSource: DataSourceOptions;
};

type AllOptions = FirestoreConnectionOptions &
  ActionOptions &
  DataSourceOptions;

/**
 * Get the configuration from a configuration file.
 */
export async function loadConfig(path?: string): Promise<AllOptions> {
  const explorer = cosmiconfig(NAME);
  const result = path ? await explorer.load(path) : await explorer.search();
  return result?.config ?? {};
}

/**
 * Returns a function that can be used to select the first
 * defined value from a list of keys from the first object.
 * If none of the keys are defined in the first object, the
 * value will be selected from the second object.
 *
 * If neither object has the keys we are looking for, an
 * empty object is returned.
 */
function greedyPick<T>(first: T, second: T) {
  return <K extends keyof T>(
    ...keys: K[]
  ): { key: K; value: T[K] } | { key?: never; value?: never } => {
    for (const key of keys) {
      if (typeof first[key] !== "undefined") {
        return { key, value: first[key] };
      }
    }
    for (const key of keys) {
      if (typeof second[key] !== "undefined") {
        return { key, value: second[key] };
      }
    }
    return {};
  };
}

/**
 * Resolve the final configuration to use from options passed through
 * the CLI, configuration file and environment variables.
 */
export async function resolveConfig(
  globalOptions: GlobalOptions,
  cliOptions: AllOptions
): Promise<ResolvedConfig> {
  const config: ResolvedConfig = {
    connection: {},
    action: {},
    dataSource: {},
  };

  const configOptions = await loadConfig(globalOptions.config);

  // Set up a selector that will pick options from the CLI first, then fallback to config file options
  const pick = greedyPick(cliOptions, configOptions);

  //
  // Connection options
  //

  const { value: project } = pick("project");
  const { key: connectionKey, value: connection } = pick(
    "emulator",
    "credentials",
    "keyFile"
  );

  if (project) config.connection.project = project;
  if (connectionKey) config.connection[connectionKey] = connection as any;

  //
  // Data options
  //

  config.action.stringify = pick("stringify").value;
  config.action.count = pick("count").value;
  config.action.limit = pick("limit").value;
  config.action.debug = pick("debug").value;
  config.action.verbose = pick("verbose").value;
  config.action.quiet = pick("quiet").value!;
  config.action.update = pick("update").value!;
  config.action.paths = pick("paths").value!;
  config.action.match = pick("match").value!;
  config.action.ignore = pick("ignore").value!;
  config.action.depth = pick("depth").value!;
  config.action.overwrite = pick("overwrite").value!;
  config.action.mode = pick("mode").value!;

  //
  // Data source options
  //

  config.dataSource.gcpProject = pick("gcpProject").value;
  config.dataSource.gcpKeyFile = pick("gcpKeyFile").value;
  config.dataSource.gcpCredentials = pick("gcpCredentials").value;
  config.dataSource.awsRegion = pick("awsRegion").value;
  config.dataSource.awsProfile = pick("awsProfile").value;
  config.dataSource.awsAccessKeyId = pick("awsAccessKeyId").value;
  config.dataSource.awsSecretAccessKey = pick("awsSecretAccessKey").value;

  return config;
}
