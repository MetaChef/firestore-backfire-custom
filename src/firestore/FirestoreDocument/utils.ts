import {
  GeoPoint,
  Query,
  DocumentReference,
  Timestamp,
} from "@google-cloud/firestore";

/**
 * Checks whether an object is a Firestore Timestamp.
 *
 * // HACK: current implementation just checks for existence of
 * the "_seconds" and "_nanoseconds" fields. Does not guarantee
 * that it is actually a Timestamp instance.
 *
 * @param object The object to check.
 * @returns `true` if the object is a Firestore Timestamp.
 */
export function isFirestoreTimestamp(object: any): object is Timestamp {
  if (typeof object === "object")
    return "_seconds" in object && "_nanoseconds" in object;
  return false;
}

/**
 * Checks whether an object is a Firestore GeoPoint.
 *
 * // HACK: current implementation just checks for existence of
 * the "_latitude" and "_longitude" fields. Does not guarantee
 * that it is actually a GeoPoint instance.
 *
 * @param object The object to check.
 * @returns `true` if the object is a Firestore GeoPoint.
 */
export function isFirestoreGeoPoint(object: any): object is GeoPoint {
  if (typeof object === "object")
    return "_latitude" in object && "_longitude" in object;
  return false;
}

/**
 * Checks whether an object is a Firestore Document Reference.
 *
 * // HACK: current implementation just checks for existence of
 * the "_firestore" and "_path" fields. Does not guarantee
 * that it is actually a DocumentReference instance.
 *
 * @param object The object to check.
 * @returns `true` if the object is a Firestore DocumentReference.
 */
export function isFirestoreDocumentReference(
  object: any,
): object is DocumentReference {
  if (typeof object === "object")
    return "_firestore" in object && "_path" in object;
  return false;
}

/**
 * Checks whether an object is a Firestore Query.
 *
 * // HACK: current implementation just checks for existence of
 * the "_firestore" and "_queryOptions" fields. Does not guarantee
 * that it is actually a Query instance.
 *
 * @param object THe object to check.
 * @returns `true` if the object is a Firestore Query.
 */
export function isFirestoreQuery(object: any): object is Query {
  if (typeof object === "object")
    return "_firestore" in object && "_queryOptions" in object;
  return false;
}
