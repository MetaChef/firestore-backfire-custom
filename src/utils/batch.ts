/**
 * Split a list into batches of the specified size.
 *
 * @param items The list of items to batch.
 * @param batchSize The size of each batch.
 */
export function splitIntoBatches<T>(items: T[], batchSize: number): T[][] {
  const batches: T[][] = [];
  for (let i = 0; i < items.length; i += batchSize) {
    batches.push(items.slice(i, i + batchSize));
  }
  return batches;
}

/**
 * Distribute a list of items evenly based on the `size` field.
 *
 * @param items The items to distribute.
 * @param numberOfBuckets The number of buckets to distribute items between.
 */
export function distributeEvenlyBySize<T extends { size: number }>(
  items: T[],
  numberOfBuckets: number
): T[][] {
  let buckets: { size: number; items: T[] }[] = [];
  for (let i = 0; i < numberOfBuckets; i++) {
    buckets.push({ size: 0, items: [] });
  }

  for (const item of items) {
    buckets[0].items.push(item);
    buckets[0].size += item.size;
    buckets = buckets.sort((a, b) => a.size - b.size);
  }

  return buckets.map((bucket) => bucket.items);
}
