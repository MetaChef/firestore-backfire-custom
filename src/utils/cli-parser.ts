import { InvalidArgumentError } from "commander";

export class CliParser {
  static integer(
    options: { min?: number; max?: number } = {}
  ): (value: string) => number {
    const { min, max } = options;
    return (value) => {
      const parsed = parseInt(value, 10);
      if (isNaN(parsed))
        throw new InvalidArgumentError(`"${value}" is not an integer.`);
      if (typeof min === "number" && parsed < min)
        throw new InvalidArgumentError(`Must be greater or equal to ${min}.`);
      if (typeof max === "number" && parsed > max)
        throw new InvalidArgumentError(`Must be less than or equal to ${max}.`);
      return parsed;
    };
  }
}
