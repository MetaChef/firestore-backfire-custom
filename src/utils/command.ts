import { InvalidArgumentError } from "commander";

/**
 * Helper function for parsing an integer and ensuring that it is
 * within the specified minimum and maximum values. If not, an
 * `InvalidArgumentError` is thrown.
 *
 * @param rawValue The raw string value to parse.
 * @param min The minimum value.
 * @param max The maximum value.
 * @returns An integer between the minimum and maximum value.
 */
export function validateMinMaxInteger(rawValue: string, min: number, max: number) {
  const value = parseInt(rawValue, 10);
  if (isNaN(value) || value < min || value > max)
    throw new InvalidArgumentError(`Must be an integer between ${min} and ${max}.`);
  return Math.max(min, Math.min(value, max));
}
