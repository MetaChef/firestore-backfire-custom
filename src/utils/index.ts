export * from "./document";
export * from "./process";
export * from "./batch";
export * from "./time";
export * from "./wait";
export * from "./regex";
