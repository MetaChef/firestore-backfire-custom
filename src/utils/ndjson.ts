/**
 * NDJSON: New-line Delimited JSON
 */
export class NDJSON {
  /**
   * Stringify a list of objects into NDJSON format.
   */
  static stringify(data: any[]): string {
    return data.map((record) => JSON.stringify(record)).join("\n");
  }

  /**
   * Parse an NDJSON string back into a list of objects.
   * Empty new-lines are skipped.
   *
   * @param options.safe If `true`, lines with parsing errors are ignored.
   */
  static parse<T = unknown>(
    text: string,
    options: { safe?: boolean } = {},
  ): T[] {
    const fail = Symbol("fail");
    return text
      .split("\n")
      .map((text) => text.trim())
      .filter((text) => text.length > 0)
      .map((text) => {
        try {
          return JSON.parse(text) as T;
        } catch (error) {
          if (options.safe) return fail;
          throw error;
        }
      })
      .filter((val): val is T => val !== fail);
  }
}
