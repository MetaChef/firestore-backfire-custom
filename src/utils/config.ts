import { cosmiconfig } from "cosmiconfig";
import parseRegex from "regex-parser";

import { Constants } from "~/config";
import { pathType } from "~/cli/helpers";
import { S3StorageSourceOptions } from "~/services/StorageSourceService/s3";
import { ExportFirestoreDataOptions, ImportFirestoreDataOptions } from "~/actions";

export async function loadConfig(path?: string) {
  const explorer = cosmiconfig(Constants.NAME);
  const result = path ? await explorer.load(path) : await explorer.search();
  return result?.config;
}

export interface GlobalOptions {
  /**
   * Path to the configuration file to use.
   */
  config?: string;
}

type Options = ExportFirestoreDataOptions | ImportFirestoreDataOptions;

export function resolveConfig<T extends Options>(
  path: string | undefined,
  commandLineOptions: Partial<T> = {},
  configOptions: Partial<T> = {}
) {
  const config: Partial<T> = {
    path: "",
    project: "",
    type: "unknown" as const,
    ...configOptions,
    ...commandLineOptions,
  };

  // Make sure command line options take priority
  if (path) config.path = path;
  if (config.path) config.type = pathType(config.path);

  const { keyfile: cliKeyfile, emulator: cliEmulator } = commandLineOptions;
  const { emulator: configEmulator } = configOptions;

  // Allow command line keyfile to take precedence over config emulator
  if (cliKeyfile && !cliEmulator && configEmulator) {
    delete config.emulator;
  }

  if (config.type === "s3") {
    const {
      awsProfile: cliProfile,
      awsAccessKeyId: cliKey,
      awsSecretAccessKey: cliSecret,
    } = commandLineOptions as S3StorageSourceOptions;

    const { awsAccessKeyId: configKey, awsSecretAccessKey: configSecret } =
      configOptions as S3StorageSourceOptions;

    // Allow command line profile to take precendence over config AWS keys
    if (cliProfile && !cliKey && !cliSecret && configKey && configSecret) {
      delete (config as S3StorageSourceOptions).awsAccessKeyId;
      delete (config as S3StorageSourceOptions).awsSecretAccessKey;
    }
  }

  // Transform path patterns from strings into RegExp
  if (config.patterns) {
    config.patterns = config.patterns.map((pattern: string | RegExp) =>
      typeof pattern === "string" ? parseRegex(pattern) : pattern
    );
  }

  return config as T;
}
