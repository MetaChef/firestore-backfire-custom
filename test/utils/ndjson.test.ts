import { describe, expect, test } from "vitest";

import { NDJSON } from "~/utils/ndjson";

describe(NDJSON.name, () => {
  describe(NDJSON.stringify.name, () => {
    test("stringifies a list of objects", () => {
      const data = [
        1,
        { hello: "world" },
        [1, 2, 3],
        [1, { hello: "world" }],
        "hello",
        "there",
      ];
      expect(NDJSON.stringify(data)).toMatchSnapshot();
    });
  });

  describe(NDJSON.parse.name, () => {
    test("parses valid NDJSON correctly", () => {
      const data = `
        {"hello":"world","list":[1,2]}
        1
        [1,2]
        "hello"
        "world"
      `;
      expect(NDJSON.parse(data)).toMatchSnapshot();
    });

    test("throws error when parsing invalid NDJSON", () => {
      const data = `
        {"hello":"world","list":[1,2]}
        [1,2
        "hello"
      `;
      expect(() => NDJSON.parse(data)).toThrowErrorMatchingSnapshot();
    });

    test("skips invalid lines of JSON when options.safe is true", () => {
      const data = `
        {"hello":"world","list":[1,2]}
        [1,2
        "hello"
      `;
      expect(NDJSON.parse(data, { safe: true })).toMatchSnapshot();
    });
  });
});
