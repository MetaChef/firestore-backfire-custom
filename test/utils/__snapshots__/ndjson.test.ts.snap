// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NDJSON > parse > parses valid NDJSON correctly 1`] = `
[
  {
    "hello": "world",
    "list": [
      1,
      2,
    ],
  },
  1,
  [
    1,
    2,
  ],
  "hello",
  "world",
]
`;

exports[`NDJSON > parse > safely ignores empty lines 1`] = `[]`;

exports[`NDJSON > parse > skips invalid lines of JSON when options.safe is true 1`] = `
[
  {
    "hello": "world",
    "list": [
      1,
      2,
    ],
  },
  "hello",
]
`;

exports[`NDJSON > parse > throws error when parsing invalid NDJSON 1`] = `"Unexpected end of JSON input"`;

exports[`NDJSON > stringify > stringifies a list of objects 1`] = `
"1
{\\"hello\\":\\"world\\"}
[1,2,3]
[1,{\\"hello\\":\\"world\\"}]
\\"hello\\"
\\"there\\""
`;
